### ISSUE TITLE
Microsoft access token with request headers

### ISSUE THREAD

**Issue by patrickconroy** [thephpleague/oauth2-client#340]
Hello. Right now the only way to get an access token from Microsoft is to add this to the `AbstractProvider::getAccessTokenRequest`, otherwise it complains that that header isn't there.

```
$options['headers']['content-type'] = 'application/x-www-form-urlencoded';
```

Can I make a PR in this repo that maybe does this, which will allow setting the headers?
```
// App..
$token = $provider->getAccessToken('authorization_code', [
    'code' => $code,
    'requestHeaders' => ['content-type' => 'application/x-www-form-urlencoded']
]);

// AbstractProvider::getAccessTokenRequest..
if (isset($params['requestHeaders'])) {
    $options['headers'] = $params['requestHeaders'];
}
return $this->getRequest($method, $url, $options);
```

Or is there a simpler way to do this? Of course I can submit a PR to the Microsoft repo, but thought I'd ask here first.

Thanks.

**Comment by rtheunissen** [comment]
I think this should be made local to the Microsoft provider, where you could override `getHeaders` to add the content type header. I'll take a closer look a bit later but that should be a good starting point.

**Comment by patrickconroy** [comment]
OK thank you @rtheunissen

**Comment by shadowhand** [comment]
@rtheunissen let's make sure that the correct content-type header is set in 1.0 and 0.1x branches.

**Comment by rtheunissen** [comment]
@shadowhand Wondering if we should defer that to Guzzle for the time being, using `form_params` instead of a string `body`? The client [sets the header for us](https://github.com/guzzle/guzzle/blob/master/src/Client.php#L294)

**Comment by shadowhand** [comment]
@rtheunissen the `form_params` option only works when you __don't__ use a pre-configured request. Our request factory doesn't support it.

**Comment by rtheunissen** [comment]
@shadowhand Good point. Should I make a plan and PR or would you like to?

**Comment by shadowhand** [comment]
I won't be able to get to it until the weekend. If you have time, go for it. :)

**Comment by SammyK** [comment]
I was just about to post a similar issue related to this. The reason behind #341 is that we're not sending an `Accept` header and Facebook is returning the wrong one by default. But Facebook returns the proper response headers when the `Accept` header is set on the request.

**Comment by moufmouf** [comment]
Hey! This is definitely not only limited to Microsoft's.
I'm currently trying to use thephpleague/oauth2-client to access an Oauth server written in PHP.

On the server side, the $_POST superglobal is empty if you do not put the correct header as suggested:

```
$options['headers']['content-type'] = 'application/x-www-form-urlencoded';
```

**Comment by rtheunissen** [comment]
@SammyK How are you currently setting the `Accept` header?

**Comment by ramsey** [comment]
Fixed in #381.
